### ISSUE TITLE
Comparison to Capn Proto

### ISSUE THREAD

**Issue by newtack** [djkoloski/rkyv#1]
rkyv looks like a very interesting project. We are using Capn Proto, but since we are developing in Rust this looks like it may be a good replacement. How does rkyv compare in terms of speed/size/security, etc. to Capn Proto?

**Comment by djkoloski** [comment]
Thanks for your interest! I'll try to address those three areas as best I can:

**Speed**
I'll preface this by saying that I have not properly benchmarked rkyv yet.
rkyv's speed should be at least comparable to Cap'n Proto, and may be better in some cases. Mostly this would be due to lower overhead traversing the data. For data of simple and intermediate complexity, I would expect both frameworks to perform the same.

**Size**
It's really tough for me to say on this one. I'm not familiar with the details of Cap'n Proto's binary format, but I do know that they claim to compress padding and unset fields. My guess is that Cap'n Proto would have the edge in size.

**Security**
It looks like Cap'n Proto has some validation on their message format, but they do also caution against handling messages from untrusted sources as they have not undergone a security review. rkyv is in the same boat, although it does not currently perform any message validation. I have some ideas floating around on how basic message validation could be introduced into rkyv.

In conclusion, I would not expect rkyv to perform much differently from Cap'n Proto. The main differences between the two frameworks will definitely be their feature sets. I would definitely encourage you to try rkyv out, find some gaps in the feature set, and file feature requests so I can evaluate what it would take to add them.

**Comment by newtack** [comment]
Thanks!

**Comment by pickfire** [comment]
Just wondering, is it possible to make rkyv support capn' proto? Like serde supporting multiple protocols.

**Comment by djkoloski** [comment]
Unfortunately it is likely impossible. rkyv is tied directly to a portable representation of types and adapting it to Cap'n Proto's format is not something that can be done. If you wanted to, you could always deserialize objects from Cap'n Proto's format into a native structure and then archive it with rkyv.
